iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergInputFormat.java

```diff
@@ -176,6 +179,20 @@ public RecordReader<Void, Container<Record>> getRecordReader(InputSplit split, J
     }
   }
 
+  private static void validateFilesWithinTableDirectory(InputSplit split, JobConf job) throws IOException {
+    boolean dataFilesWithingTableLocationOnly =
+        job.getBoolean(HiveConf.ConfVars.HIVE_ICEBERG_ALLOW_DATA_IN_TABLE_LOCATION_ONLY.varname,
+            HiveConf.ConfVars.HIVE_ICEBERG_ALLOW_DATA_IN_TABLE_LOCATION_ONLY.defaultBoolVal);
+    if (dataFilesWithingTableLocationOnly) {
+      Path tableLocation = new Path(job.get(InputFormatConfig.TABLE_LOCATION));
```

Comment:
why do we restrict on reads - that is expensive? we shouldn't allow any additions outside of table dir

Reply:
the read path constraint is to avoid to read other location's - aka other table's - data from such a malicious table. Such table can be constructed manually, not necessarily written by spark (actually most probably constructed with other methods). 
The problem here is that without this read limitation, the user can use hive's elevated privileges (doAs=false) to access secured data even if data doesn't belong to the user's own table.

Reply:
Users can also place a custom jar into the classpath to bypass that check. It doesn't seem like a robust solution, but more like a hack.

Reply:
> Would a custom/per-user jar could lead to the same class override?
yes, but that could be altered only by admin

> Why don't we consider storage-based authorization
why can't we use `doAs=true` as a quick fix?

> not expose the metadata
if the data location is sensitive - why should we leak it? Also, this way authorization would be done in 1 dedicated place (service designed for that), but not in every engine.

Reply:
If jar injection to override behaviour is only possible by admin, then this should not be a blocker for this scope. E.g. admin could even inject jar to override the AuthN chain and dump username/password pairs from those who are connecting to HS2 e.g. via jdbc via this auth method.

doAs=true is not an option for Fine-Grained Access Control where this issues is the most significant (e.g. otherwise masked data breached as non-masked). Spark has no FGAC and elevated privilege based data access decoupling from end-user based file-access. 

> if the data location is sensitive - why should we leak it?

Historically it was not sensitive and even with Iceberg it should not be treated as sensitive. The issues comes from  Iceberg's new behaviour where instead of limiting the read to a directory in a Hive table format case, Iceberg now can read data files from anywhere the hive service user has access to, if the location is in its manifest file.

Reply:
I agree, and it is also breaking the multi-data locations in case of tiered-storage usage or just the movement of the Iceberg table's _write.data.path_, but that's why this is behind a configuration flag not enabled by default and considered only as a temporary quick fix for those who are rather break it temporarily (especially if they use Iceberg as standard external table only not noticing the limitations), than experience data breach. We are also already discussing a more complex possible solution where neither the Iceberg API+functionality are limited, nor the malicious tables could expose other tables' data.

Reply:
ok, also one of the findings (Implement LOAD data for partitioned tables via Append API) won't be possible under this config

Reply:
Right. We can articulate the limitation better in the HiveConf.java change, like adding that this breaks all Iceberg tables with data located outside of the table's location.